csv/src/fs2/data/csv/CellDecoder.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2019 Lucas Satabin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package fs2.data.csv
+
+import cats._
+import cats.implicits._
+
+import scala.util.Try
+import scala.annotation.tailrec
+
+/** Describes how a cell can be decoded to the given type.
+  */
+trait CellDecoder[T] {
+  def apply(cell: String): DecoderResult[T]
+}
+
+object CellDecoder {
+
+  implicit object CellDecoderMonadError extends MonadError[CellDecoder, DecoderError] {
+    def flatMap[A, B](fa: CellDecoder[A])(f: A => CellDecoder[B]): CellDecoder[B] =
+      s => fa(s).flatMap(f(_)(s))
+
+    def handleErrorWith[A](fa: CellDecoder[A])(f: DecoderError => CellDecoder[A]): CellDecoder[A] =
+      s => fa(s).leftFlatMap(f(_)(s))
+
+    def pure[A](x: A): CellDecoder[A] =
+      s => Right(x)
+
+    def raiseError[A](e: DecoderError): CellDecoder[A] =
+      s => Left(e)
+
+    def tailRecM[A, B](a: A)(f: A => CellDecoder[Either[A, B]]): CellDecoder[B] = {
+      @tailrec
+      def step(s: String, a: A): DecoderResult[B] =
+        f(a)(s) match {
+          case left @ Left(_)          => left.asInstanceOf[DecoderResult[B]] // avoid creating too many objects
```

Comment:
cats has some nice `withRight` and `withLeft` helpers for exactly these casts.
```suggestion
          case left @ Left(_)          => left.withRight[B] // avoid creating too many objects
```

Reply:
`withRight` expects `B` to be a super type of whatever right type is for `left`. In this case it is an `Either[A, B]`, so it doesn't work, I need `rightCast`.